MaterialLib/MPL/Properties/RelPermLiakopoulos.cpp

```diff
@@ -0,0 +1,105 @@
+/**
+ * \file
+ * \author Norbert Grunwald
+ * \brief
+ *
+ * \copyright
+ * Copyright (c) 2012-2018, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include "MaterialLib/MPL/Properties/RelPermLiakopoulos.h"
+#include "MaterialLib/MPL/Medium.h"
+
+#include <algorithm>
+#include <cmath>
+
+namespace MaterialPropertyLib
+{
+/**
+ */
+PropertyDataType RelPermLiakopoulos::value(
+    VariableArray const& variable_array,
+    ParameterLib::SpatialPosition const& pos,
+    double const t) const
+{
+    /// here, an extra computation of saturation is forced, guaranteeing a
+    /// correct value. In order to speed up the computing time, saturation could
+    /// be insertred into the primary variable array after it is computed in the
+    /// FEM assembly.
+    auto const s_L = _medium->property(PropertyType::saturation)
+                         .template value<double>(variable_array, pos, t);
+    auto const s_L_res = _residual_liquid_saturation;
+    auto const k_rel_min_GR = _min_relative_permeability_gas;
+
+    if (s_L <= s_L_res)
+    {
+        return Pair{0., 1.};
+    }
+
+    if (s_L >= 1.)
+    {
+        return Pair{1., k_rel_min_GR};
+    }
+
+    auto const a = _parameter_a;
+    auto const b = _parameter_b;
+    auto const lambda = _exponent;
+
+    auto const s_eff = (s_L - s_L_res) / (1. - s_L_res);
+
+    auto const k_rel_LR = 1. - a * std::pow(1. - s_L, b);
+    auto const k_rel_GR = (1. - s_eff) * (1. - s_eff) *
+                          (1. - std::pow(s_eff, (2. + lambda) / lambda));
+
+    const Pair kRel = {std::max(k_rel_LR, 0.),
+                       std::max(k_rel_GR, k_rel_min_GR)};
+
+    return kRel;
+}
+PropertyDataType RelPermLiakopoulos::dValue(
+    VariableArray const& variable_array, Variable const primary_variable,
+    ParameterLib::SpatialPosition const& pos, double const t) const
+{
+    (void)primary_variable;
+    assert((primary_variable == Variable::liquid_saturation) &&
+           "RelPermLiakopoulos::dValue is implemented for "
+           " derivatives with respect to liquid saturation only.");
+    auto const s_L = _medium->property(PropertyType::saturation)
+                         .template value<double>(variable_array, pos, t);
+    auto const s_L_res = _residual_liquid_saturation;
+    if (s_L < s_L_res)
```

Comment:
Suggestion to restrict the function within [s_r, s_max]. For example,
```suggestion
   const double s_within_range = std::min(std::max(s_L_res, s_L), 1.0); //or replace 1.0 with s_L_max
```
Then replace `s_L` with ` s_within_range` in the following lines.

Reply:
OK, i do so. However, this affects the dry condition only, which is not considered in the Liakopoulos experiment (Saturation does not drop below .7 or so). Anyway, this is a better approach.